godot-core/src/builtin/signal/futures.rs

```diff
@@ -0,0 +1,313 @@
+/// [`Result::Err`]
+pub struct SignalFuture<R: ParamTuple + Sync + Send>(TrySignalFuture<R>);
+
+impl<R: ParamTuple + Sync + Send> SignalFuture<R> {
+    fn new(signal: Signal) -> Self {
+        Self(TrySignalFuture::new(signal))
+    }
+}
+
+impl<R: ParamTuple + Sync + Send> Future for SignalFuture<R> {
+    type Output = R;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let poll_result = self.get_mut().0.poll(cx);
+
+        match poll_result {
+            Poll::Pending => Poll::Pending,
+            Poll::Ready(Ok(value)) => Poll::Ready(value),
+            Poll::Ready(Err(TrySignalFutureError)) => panic!(
+                "The signal object of a SignalFuture was freed while the future was still waiting for the signal to be emitted!"
+            ),
+        }
+    }
+}
+
+// Not derived, otherwise an extra bound `Output: Default` is required.
+struct SignalFutureData<T> {
+    state: TrySignalFutureState<T>,
+    waker: Option<Waker>,
+}
+
+impl<T> Default for SignalFutureData<T> {
+    fn default() -> Self {
+        Self {
+            state: Default::default(),
+            waker: None,
+        }
+    }
+}
+
+// Only public for itest.
+#[cfg_attr(feature = "trace", derive(Default))]
+pub struct SignalFutureResolver<R> {
+    data: Arc<Mutex<SignalFutureData<R>>>,
+}
+
+impl<R> Clone for SignalFutureResolver<R> {
+    fn clone(&self) -> Self {
+        Self {
+            data: self.data.clone(),
+        }
+    }
+}
+
+impl<R> SignalFutureResolver<R> {
+    fn new(data: Arc<Mutex<SignalFutureData<R>>>) -> Self {
+        Self { data }
+    }
+}
+
+impl<R> std::hash::Hash for SignalFutureResolver<R> {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        state.write_usize(Arc::as_ptr(&self.data) as usize);
+    }
+}
+
+impl<R> PartialEq for SignalFutureResolver<R> {
+    fn eq(&self, other: &Self) -> bool {
+        Arc::ptr_eq(&self.data, &other.data)
+    }
+}
+
+impl<R: ParamTuple + Sync + Send> RustCallable for SignalFutureResolver<R> {
+    fn invoke(&mut self, args: &[&Variant]) -> Result<Variant, ()> {
+        let mut data = self.data.lock().unwrap();
+        // We no loger need the waker after we resolved. If the future get's polled again we also get a new waker.
+        let waker = data.waker.take();
+
+        data.state = TrySignalFutureState::Ready(R::from_variant_array(args));
+        drop(data);
+
+        if let Some(waker) = waker {
+            waker.wake();
+        }
+
+        Ok(Variant::nil())
+    }
+}
+
+impl<R> Display for SignalFutureResolver<R> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "GuaranteedSignalFutureResolver::<{}>",
+            std::any::type_name::<R>()
+        )
+    }
+}
+
+// this resolver will resolve the future when it's being dropped (i.e. the engine removes all connected signal callables). This is very unusual.
```

Comment:
```suggestion
// This resolver will resolve the future when it's being dropped (i.e. the engine removes all connected signal callables). This is very unusual.
```

"Unusual" as in "unexpected" (i.e. bad), or what does that mean?

Reply:
This was related to the unusual behavior of the `GuaranteedFuture` and shouldn't be here anymore. I'm removing it.